importer/balanced/builder.go

```diff
@@ -27,87 +43,114 @@ import (
-			if err := nroot.AddChild(root, db); err != nil {
-				return nil, err
-			}
-		}
-
-		// fill it up.
-		if err := fillNodeRec(db, nroot, level, offset); err != nil {
-			return nil, err
-		}
-
-		offset = nroot.FileSize()
-		root = nroot
-
-	}
-	if root == nil {
-		// this should only happen with an empty node, so return a leaf
-		var err error
-		root, err = db.NewLeaf(nil)
+	if db.Done() {
+		// No data, return just an empty node.
+		root, err := db.NewLeafNode(nil)
 		if err != nil {
 			return nil, err
 		}
+		// This works without Filestore support (`ProcessFileStore`).
+		// TODO: Why? Is there a test case missing?
+
+		return db.AddNodeAndClose(root)
 	}
 
-	out, err := db.Add(root)
+	// The first `root` will be a single leaf node with data
+	// (corner case), after that subsequent `root` nodes will
+	// always be internal nodes (with a depth > 0) that can
+	// be handled by the loop.
+	root, fileSize, err := db.NewLeafDataNode()
 	if err != nil {
 		return nil, err
 	}
 
-	err = db.Close()
-	if err != nil {
-		return nil, err
+	// Each time a DAG of a certain `depth` is filled (because it
+	// has reached its maximum capacity of `db.Maxlinks()` per node)
+	// extend it by making it a sub-DAG of a bigger DAG with `depth+1`.
+	for depth := 1; !db.Done(); depth++ {
+
+		// Add the old `root` as a child of the `newRoot`.
+		newRoot := db.NewFSNodeOverDag(ft.TFile)
+		newRoot.AddChild(root, fileSize, db)
+
+		// Fill the `newRoot` (that has the old `root` already as child)
+		// and make it the current `root` for the next iteration (when
+		// it will become "old").
+		root, fileSize, err = fillNodeRec(db, newRoot, depth)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return out, nil
+	return db.AddNodeAndClose(root)
 }
 
-// fillNodeRec will fill the given node with data from the dagBuilders input
-// source down to an indirection depth as specified by 'depth'
-// it returns the total dataSize of the node, and a potential error
+// fillNodeRec will fill the given internal (non-leaf) `node` with either
+// more internal nodes or leaf data nodes (when the final `depth` of 1 is
+// reached) containing the file data passed from the `DagBuilderHelper`.
+//
+// If `node` is `nil` it means we're descending into new territory so
+// just create a new node to be filled (recursion call); if it's not
+// `nil` it has been called from `Layout` where the DAG has been extended
+// to a new root, that already has a child (the old root), to be filled
+// with more children (up to `db.Maxlinks()`).
+//
+// The `node` to be filled uses the `FSNodeOverDag` abstraction that
+// allows to add child nodes without packing/unpacking the UnixFS layer
+// node (having an internal `ft.FSNode` cache).
+//
+// It returns the `ipld.Node` representation of the passed `node` filled
+// with children and the `nodeFileSize` with the size of the file stored
+// under this node (to propagate it upwards in the DAG to its parent).
```

Comment:
Instead of:

> `nodeFileSize` with the size of the file stored under this node

maybe:

> `nodeFileSize` with the total size of all nodes under this node

or 

> `nodeFileSize` with the total size of the file chunk (leaf) nodes stored under this node

“size of the file” makes it unclear whether it’s the *whole* file or just the relevant chunks.